### PR TITLE
Add optional json output from test runner

### PR DESCRIPTION
**Pull request by rundis**
**TLDR:** This pull request add support for outputting test results as JSON. You can now get results in json format as follows:
`elm-test --report=json`

## Background
For tooling purposes (my angle is editor integration) it would be very useful to have a more structured format for integrating with elm-test. JSON seems like a suitable place to start and should hopefully be useful for most editors and other tools.

## Solution overview
screen shot 2016-10-09 at 15 47 45

- Refactored out the existing Chalk related reporting into a separate module
- Added a JSon reporter module
- Added a module for shared types between the two `Test.Reporter.Result` (might be overkill ?) 
- Added a module `Test.Reporter.Reporter` which creates a `TestReporter` record to try to encapsulate the Node runner from dealing with which format is being used

## Testing
I wasn't able to get the bundled `tests\ci.js`to work properly on my machine, not sure what I'm doing wrong. I'm happy to try to get some sort of testing running for the new stuff if you think it's necessary (which I'm guessing the answer to would be yes). I have tested manually on the command-line with a couple of projects and have successfully integrated with the runner from Light Table.

## Other improvements
- Related issue in elm-test: https://github.com/elm-community/elm-test/issues/74

### CODE REVIEW

**Comment by rundis** [comment]
@rtfeldman Happy to make any changes to this PR as you see fit. I also noticed that the builds have failed prior to this PR so I guess I can't necessarily be blamed for the build of this PR failing in travis.

**Comment by rtfeldman** [comment]
Looks great, thanks @rundis!
